### ISSUE TITLE
UnnecessaryNullCheckWithInstanceOf ignores redundant null check for complex cases.

### ISSUE THREAD

**Issue by Anmol202005** [checkstyle/checkstyle#17137]
I have read check documentation: https://checkstyle.org/checks/coding/unnecessarynullcheckwithinstanceof.html#UnnecessaryNullCheckWithInstanceOf
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

Follow up of : https://github.com/checkstyle/checkstyle/pull/16489

```bash
/var/tmp $ javac Test.java

/var/tmp $ cat config.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
        "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
  <module name="TreeWalker">
    <module name="UnnecessaryNullCheckWithInstanceOf"/>
  </module>
</module>
```
```bash
/var/tmp $ cat Test.java
package com.puppycrawl.tools.checkstyle;

import java.util.Objects;

public class Test {
    public void processData(Object data, String s, String b, String getData) {
        boolean result = false;
        while (getData!= null && getData instanceof String) {
           result = true;
        }
        if (data != null && (!Objects.equals(s, b) || data instanceof String)) {  // false-negative
            boolean isValid = b != null && (s != null || b instanceof String);    //  false-negative
        }

    }
}


/var/tmp $ RUN_LOCALE="-Duser.language=en -Duser.country=US"
/var/tmp $ java $RUN_LOCALE -jar checkstyle-10.24.0-SNAPSHOT-all.jar -c config.xml Test.java
Starting audit...
[ERROR] C:\Users\anmol\IdeaProjects\checkstyle3\src\main\java\com\puppycrawl\tools\checkstyle\Test.java:8:16: Unnecessary nullity check with instanceof operator in conditional expression. [UnnecessaryNullCheckWithInstanceOf]
Audit done.
Checkstyle ends with 1 errors.

```

the null comparison in redundant in the expression `data != null && (!Objects.equals(s, b) || data instanceof String` and `boolean isValid = b != null && (s != null || b instanceof String`  therefore a violation is expected at line 11 and 12.

---

**Comment by StefanLapointe** [comment]
The null check in `data != null && (!Objects.equals(s, b) || data instanceof String)` is not redundant. It is impossible to distinguish between cases where data is null and cases where data is a non-String object using only a logical expression of only `Objects.equals(s, b)` and `data instanceof String)`. A similar argument applies to the other example.

I think perhaps this issue should be closed, unless there is still something still wrong with this check.

As for complex cases in general, I think that it is probably best to only handle cases where `obj != null` and `obj instanceof Foo` are purely being `&&`'d together with possibly other stuff, since this is the only case where the instanceof expression automatically makes the expression false in the null case even with the null check removed. Once you start mixing in `||`, null checks can become non-redundant. I think the current implementation is satisfactory as-is in this regard.

Maybe we should add support for the `&` operator though. It can be used for logic even though its primary purpose is a bitwise operation. It doesn't short-circuit when the first operand is `false` like `&&` does. We would need to replace the expression `nullCheckNode == null && currentChild.getType() == TokenTypes.LAND` in the method `findRedundantNullCheck()` with something that also checks for `TokenTypes.BAND`.

**Comment by Atharv3221** [comment]
I am working on this issue

**Comment by romani** [comment]
fix from @MonuChaudhary14  is merged.
